drivers/timer/mchp_xec_rtos_timer.c

```diff
@@ -0,0 +1,271 @@
+}
+
+/* SYSTICK elapsed function */
+#if 0
+static u32_t elapsed(void)
+{
+	u32_t val, ctrl1, ctrl2;
+
+	ctrl1 = GIRQ23_REGS->SRC;
+	val = RTMR_REGS->CNT & COUNTER_MAX;
+	ctrl2 = GIRQ23_REGS->SRC;
+
+	overflow_cyc += (ctrl1 & MCHP_RTMR_GIRQ_VAL) ? last_load : 0;
+	if (val > VAL_ABOUT_TO_WRAP) {
+		int wrap = ctrl2 & MCHP_RTMR_GIRQ_VAL;
+
+		overflow_cyc += (wrap != 0) ? last_load : 0;
+	}
+
+	return (last_load - val) + overflow_cyc;
+}
+#endif
+
+/*
+ * RTOS Timer runs at 32KHz and counts down to 0. The number of counts
+ * elpapsed since the timer was started is Preload register - Count register.
+ * This routine is expected to return the number of ticks elapsed since the
+ * last time this routine was called. This implementation depends upon
+ * one-shot behaviour (auto-reload NOT enabled).
+ */
+static u32_t elapsed(void)
+{
+	u32_t tdiff = (last_load - (RTMR_REGS->CNT & COUNTER_MAX));
+	u32_t elapsed = 0;
+
+	if (tdiff != elapsed_counts) {
+		elapsed_counts = tdiff;
+		elapsed = tdiff;
+	}
+	return elapsed;
+}
+
+void xec_rtos_timer_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+	u32_t dticks;
+
+#ifdef DEBUG_RTOS_TIMER
+	GPIO_CTRL_REGS->CTRL_0015 ^= (1ul << 16);
+#endif
+
+	timer_clear_intr();
+
+	cycle_count += last_load;
+	dticks = (cycle_count - announced_cycles) / CYC_PER_TICK;
+	announced_cycles += dticks * CYC_PER_TICK;
+
+	z_clock_announce(TICKLESS ? dticks : 1);
+}
+
+int z_clock_driver_init(struct device *device)
+{
+	mchp_pcr_periph_slp_ctrl(PCR_RTMR, 0);
+	mchp_pcr_periph_reset(PCR_RTMR);
+	timer_clear_intr();
+
+	last_load = CYC_PER_TICK;
+	cycle_count = 0U;
+	announced_cycles = 0U;
+	elapsed_counts = 0U;
+
+	IRQ_CONNECT(RTMR_IRQn, 1, xec_rtos_timer_isr, 0, 0);
+	irq_enable(RTMR_IRQn);
+
+	timer_preload_set(CYC_PER_TICK);
+	timer_counter_set(CYC_PER_TICK);
+
+	timer_ien();
+	timer_start();
+
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	/* Fast CPUs and a 24 bit counter mean that even idle systems
+	 * need to wake up multiple times per second.  If the kernel
+	 * allows us to miss tick announcements in idle, then shut off
+	 * the counter. (Note: we can assume if idle==true that
+	 * interrupts are already disabled)
+	 */
+	if (IS_ENABLED(CONFIG_TICKLESS_IDLE) && idle && ticks == K_FOREVER) {
+		timer_pause();
+		return;
+	}
+
+#if defined(CONFIG_TICKLESS_KERNEL) && !defined(CONFIG_QEMU_TICKLESS_WORKAROUND)
+	u32_t delay;
+
+	ticks = MIN(MAX_TICKS, MAX(ticks - 1, 0));
+
+	/* Desired delay in the future */
+	delay = (ticks == 0) ? MIN_DELAY : ticks * CYC_PER_TICK;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	cycle_count += elapsed();
+
+	/* Round delay up to next tick boundary */
+	delay = delay + (cycle_count - announced_cycles);
+	delay = ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
+	last_load = delay - (cycle_count - announced_cycles);
+
+	timer_stop();
+	delay = last_load;
+	timer_clear_intr();
+	timer_preload_set(delay);
+	timer_counter_set(delay);
+	timer_start();
+
```

Comment:
Does the hardware really require you do this dance every time you want to change the timeout value?  That's... sort of unfortunate.   This call gets made often, sometimes with tight latency requirements, and we really don't want it halting the counter.

Think about it: one of the worst side effects imaginable of code that spins managing a zillion small timeouts is that the underlying clock starts to slow down!

Other architectures all allow the counter to be set while it's ticking.  My hope is this one does too.

Reply:
Yes, the HW design requires this.  When I didn't do it the timer would not take the new preload or count register values and wrap from 0.  I have confirmed with the designers that there is a race condition that requires disable timer, load count and/or preload, enable timer.

Reply:
Yeah, then that definitely deserves a comment to prevent people like me from coming along and trying to "fix" it. :)